ami/socket.go

```diff
@@ -31,6 +32,7 @@ func NewSocket(ctx context.Context, address string) (*Socket, error) {
 		conn:     conn,
 		incoming: make(chan string, 32),
 		shutdown: make(chan struct{}),
+		error:    make(chan error),
```

Comment:
You are storing the errors into this channel, but don't returning them.
Suggestion:
 - change the name to errors
 -  in the close method, returns the err like:

```go
func (s *Socket) Close(ctx context.Context) error {
...
select {
case err := <- s.errors:
  s.terminate()
  return err
case <-done:
case <-time.NewTimer(150 * time.Millisecond).C:
case <- ctx.Done():
}
return s.terminate()
}
```

Please execute the unit tests before submit the patch

Reply:
Hi, i'm returning the error in `func Recv`.
the `s.terminate()` just closes the connections but do nothing with commands waiting for responses (
the idea behind the "error" channel is to prevent `func send` get stuck waiting for response.
Hope i've explained (not english speaker)